django_afip/models.py

```diff
@@ -944,41 +945,44 @@ def validate(self, ticket: AuthTicket | None = None) -> list[str]:
             return []
 
         qs.order_by("issued_date", "id")._assign_numbers()
+        ids = qs.values_list("id", flat=True)
 
-        ticket = ticket or first.point_of_sales.owner.get_or_create_ticket("wsfe")
-        client = clients.get_client("wsfe", first.point_of_sales.owner.is_sandboxed)
-        response = client.service.FECAESolicitar(
-            serializers.serialize_ticket(ticket),
-            serializers.serialize_multiple_receipts(qs),
-        )
-        check_response(response)
         errs = []
-        for cae_data in response.FeDetResp.FECAEDetResponse:
-            if cae_data.Resultado == ReceiptValidation.RESULT_APPROVED:
-                validation = ReceiptValidation.objects.create(
-                    result=cae_data.Resultado,
-                    cae=cae_data.CAE,
-                    cae_expiration=parsers.parse_date(cae_data.CAEFchVto),
-                    receipt=qs.get(
-                        receipt_number=cae_data.CbteDesde,
-                    ),
-                    processed_date=parsers.parse_datetime(
-                        response.FeCabResp.FchProceso,
-                    ),
-                )
-                if cae_data.Observaciones:
+        with transaction.atomic():
+            qs = Receipt.objects.filter(id__in=ids, receipt_number__isnull=False).select_for_update()
```

Comment:
Podés reusar el QuerySet existente acá:

```suggestion
            qs = qs.select_for_update()
```

Reply:
Eso intenté en un principio, pero me da esta excepción:

```python
self = <django.db.backends.utils.CursorWrapper object at 0x7f219c6762a0>
sql = 'SELECT "afip_receipt"."id", "afip_receipt"."point_of_sales_id", "afip_receipt"."receipt_type_id", "afip_receipt"."con..."."id" = %s AND "afip_receiptvalidation"."id" IS NULL) ORDER BY "afip_receipt"."receipt_number" ASC LIMIT 1 FOR UPDATE'
params = (12,), ignored_wrapper_args = (False, {'connection': <DatabaseWrapper vendor='postgresql' alias='default'>, 'cursor': <django.db.backends.utils.CursorWrapper object at 0x7f219c6762a0>})

    def _execute(self, sql, params, *ignored_wrapper_args):
        # Raise a warning during app initialization (stored_app_configs is only
        # ever set during testing).
        if not apps.ready and not apps.stored_app_configs:
            warnings.warn(self.APPS_NOT_READY_WARNING_MSG, category=RuntimeWarning)
        self.db.validate_no_broken_transaction()
        with self.db.wrap_database_errors:
            if params is None:
                # params default might be backend specific.
                return self.cursor.execute(sql)
            else:
>               return self.cursor.execute(sql, params)
E               django.db.utils.NotSupportedError: FOR UPDATE cannot be applied to the nullable side of an outer join

.tox/live/lib/python3.12/site-packages/django/db/backends/utils.py:105: NotSupportedError

```

Me parece que el problema es el `qs = self.filter(validation__isnull=True)` de un par de lineas más arriba. Por lo que leí, Postgres dice que no puede lockear las filas según un criterio como ese, porque en cualquier momento podría suceder que `validation__isnull=True` si se agregara un Validation que apunte al `Receipt`; tendría que lockear la tabla Validations entera supongo, o algo similar.

Decime si estoy tirando fruta, no sé mucho PG pero menos aun Django